socorro/processor/signature_utilities.py

```diff
@@ -93,6 +93,63 @@ def __init__(self, config, quit_check_callback=None):
         self.fixup_comma = re.compile(r',(?! )')
 
     #--------------------------------------------------------------------------
+    def _is_exception(
+        self,
+        exception_list,
+        remaining_original_line,
+        line_up_to_current_position
+    ):
+        for an_exception in exception_list:
+            if remaining_original_line.startswith(an_exception):
+                return True
+            if line_up_to_current_position.endswith(an_exception):
+                return True
+        return False
+
+    #--------------------------------------------------------------------------
+    def _collapse(
+        self,
+        function,
+        open_string,
+        replacement_open_string,
+        close_string,
+        replacement_close_string,
+        exception_substring_list=(),  # list of exceptions that shouldn't collapse
+    ):
+        target_counter = 0
+        collapsed_list = []
+        exception_mode = False
+
+        def append_if_not_in_collapse_mode(a_character):
+            if not target_counter:
+                collapsed_list.append(a_character)
+
+        for index, a_character in enumerate(function):
+            if a_character == open_string:
+                if self._is_exception(
+                    exception_substring_list,
+                    function[index + 1:],
+                    function[:index]
+                ):
+                    exception_mode = True
+                    append_if_not_in_collapse_mode(a_character)
+                    continue
+                append_if_not_in_collapse_mode(replacement_open_string)
+                target_counter += 1
+            elif a_character == close_string:
+                if exception_mode:
+                    append_if_not_in_collapse_mode(a_character)
+                    exception_mode = False
+                else:
+                    target_counter -= 1
+                    append_if_not_in_collapse_mode(replacement_close_string)
+            else:
+                append_if_not_in_collapse_mode(a_character)
+
+        edited_function = ''.join(collapsed_list)
+        return edited_function
+
```

Comment:
I can't put my finger on exactly where it goes wrong but this code doesn't protect against unbalanced `open_string` and `close_string`. 
E.g. `_collapse('less<more', '<', '{', '>', 'T}') == 'less{'`
Is that a problem?